File: 卡码/代码随想录/1/10-单调栈/Go/LeetCode_0496.go
// 496. 下一个更大元素 I
package main

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	res := make([]int, len(nums1))
	for i := 0; i < len(nums1); i++ {
		res[i] = -1
	}
	st := make([]int, 0)
	m := make(map[int]int, 0)
	for index, num := range nums1 {
		m[num] = index
	}
	for index, num := range nums2 {
		for len(st) > 0 && num > nums2[st[len(st)-1]] {
			if _, ok := m[nums2[st[len(st)-1]]]; ok {
				res[m[nums2[st[len(st)-1]]]] = num
			}
			st = st[:len(st)-1]
		}
		st = append(st, index)
	}
	return res
}

File: 卡码/代码随想录/1/10-单调栈/Js/LeetCode_0496.js
//496. 下一个更大元素 I
const nextGreaterElement = function(nums1, nums2) {
	const res=new Array(nums1.length).fill(-1);
	const st=new Array();
	const hash=new Map();
	for(let i=0;i<nums1.length;i++) hash.set(nums1[i],i);
	for(let i=0;i<nums2.length;i++){
		while(st.length&&nums2[i]>nums2[st[st.length-1]]){
			if(hash.has(nums2[st[st.length-1]])) res[hash.get(nums2[st[st.length-1]])]=nums2[i];
			st.pop();
		}
		st.push(i);
	}
	return res;
};
